tests/base/src/test/java/org/wildfly/security/ssl/SSLv2HelloAuthenticationTest.java

```diff
@@ -0,0 +1,479 @@
+
+        assertNotNull(identity);
+        assertEquals("Principal Name", "ladybird", identity.getPrincipal().getName());
+    }
+
+    /**
+     * Test two way authentication when the client does not support SSLv2Hello,
+     * but the server has SSLv2Hello support enabled.
+     */
+    @Test
+    public void testTwoWaySSLv2HelloNoClientSupport() throws Exception {
+        List<Protocol> list = new ArrayList<>();
+        list.add(Protocol.forName("SSLv2Hello"));
+        list.add(Protocol.forName("TLSv1"));
+
+        SSLContext serverContext = new SSLContextBuilder()
+                .setSecurityDomain(securityDomain)
+                .setKeyManager(getKeyManager("/ca/jks/scarab.keystore"))
+                .setTrustManager(getCATrustManager())
+                .setNeedClientAuth(true)
+                .setProtocolSelector(ProtocolSelector.empty().add(EnumSet.copyOf(list)))
+                .build().create();
+
+        // For IBM JDK, although the server accepts SSLv2Hello messages, it will not display it in its list
+        // of enabled protocols.
+        String[] enabledServerProtocols = IS_IBM ? new String[]{"TLSv1"} : new String[]{"SSLv2Hello", "TLSv1"} ;
+        String[] enabledClientProtocols = new String[]{"TLSv1", "TLSv1.1"};
+
+        SecurityIdentity identity = performConnectionTest(serverContext,
+                "protocol://two-way-no-sslv2hello.org",
+                "wildfly-ssl-test-config-v1_6.xml",
+                enabledClientProtocols,
+                enabledServerProtocols,
+                "TLSv1"); // We expect the negotiated protocol to be the highest version protocol in common
+        assertNotNull(identity);
+        assertEquals("Principal Name", "ladybird", identity.getPrincipal().getName());
+    }
+
+    /**
+     * Test two way authentication where the client supports SSLv2Hello but the server does not.
+     * Handshake should fail.
+     */
+    @Test
+    public void testTwoWaySSlv2HelloNoServerSupport() throws Exception {
+
+        Assume.assumeFalse("Skipping testTwoWaySSLv2HelloNoServerSupport test " +
+                "as IBM JDK does not support SSLv2Hello on the " +
+                "client side", IS_IBM);
+
+            List<Protocol> list = new ArrayList<>();
+            list.add(Protocol.forName("TLSv1.1"));
+
+            SSLContext serverContext = new SSLContextBuilder()
+                    .setSecurityDomain(securityDomain)
+                    .setKeyManager(getKeyManager("/ca/jks/scarab.keystore"))
+                    .setTrustManager(getCATrustManager())
+                    .setNeedClientAuth(true)
+                    .setProtocolSelector(ProtocolSelector.empty().add(EnumSet.copyOf(list)))
+                    .build().create();
+
+            String[] serverEnabledProtocols = new String[]{"TLSv1.1"};
+            String[] clientEnabledProtocols = new String[]{ "SSLv2Hello", "TLSv1"};
+
+            SecurityIdentity identity = performConnectionTest(serverContext,
```

Comment:
Would be good to assert that `identity` is `null`.

Reply:
Added